cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogFolderCleaner.java

```diff
@@ -0,0 +1,110 @@
+import java.nio.file.SimpleFileVisitor;
+
+/**
+ * LogFolderCleaner is responsible for cleaning up accumulated empty logs folders.
+ */
+public class LogFolderCleaner implements Runnable {
+
+    /**
+     * FolderCleanupFileVisitor traverses the file tree in Depth-First manner.
+     * Cleanups up empty folders.
+     * @param <T>
+     */
+    private class FolderCleanupFileVisitor<T extends Path> extends SimpleFileVisitor<Path> {
+        private final int folderCleanupBatchSize;
+        private final long sleepDurationMs;
+
+        private int count;
+
+        /**
+         * Parameterized constructor for FolderCleanupFileVisitor
+         * @param folderCleanupBatchSize Maximum number of folders to clean before sleeping
+         * @param sleepDurationMs Sleep time in milliseconds after cleaning up a batch of folders
+         */
+        protected FolderCleanupFileVisitor(int folderCleanupBatchSize, long sleepDurationMs) {
+            this.folderCleanupBatchSize = folderCleanupBatchSize;
+            this.sleepDurationMs = sleepDurationMs;
+            this.count = 0;
+        }
+
+        @Override
+        public FileVisitResult postVisitDirectory(final Path dir, IOException e) throws IOException {
+            if (isDirectoryEmpty(dir)) {
+                Files.deleteIfExists(dir);
+            }
+
+            count++;
+            if (count == folderCleanupBatchSize) {
+                count = 0;
+                try {
+                    Thread.sleep(sleepDurationMs);
+                } catch (InterruptedException ex) {
+                    LOG.warn("InterruptedException exception occurred.", ex);
+                }
+            }
+            return FileVisitResult.CONTINUE;
+        }
+
+        private boolean isDirectoryEmpty(final Path directory) throws IOException {
+            try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(directory)) {
+                return !dirStream.iterator().hasNext();
+            }
+        }
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(LogFolderCleaner.class);
+
+    private final Location logsDirectoryLocation;
+    private final int folderCleanupBatchSize;
+    private final long sleepDurationMs;
+
+    /**
+     * Parameterized constructor for LogFolderCleaner
+     * @param logsDirectoryLocation Logs parent folder location
+     * @param folderCleanupBatchSize Maximum number of folders to clean before sleeping
+     * @param sleepDurationMs Sleep time in milliseconds after cleaning up a batch of folders
+     */
+    public LogFolderCleaner(Location logsDirectoryLocation, int folderCleanupBatchSize, long sleepDurationMs) {
+        this.logsDirectoryLocation = logsDirectoryLocation;
+        this.folderCleanupBatchSize = folderCleanupBatchSize;
+        this.sleepDurationMs = sleepDurationMs;
+    }
+
+    @Override
+    public void run() {
+        try {
+            Path logFolderPath = Paths.get(logsDirectoryLocation.toURI());
```

Comment:
Any specific reason to move to files only here? As soon as we want everything to be location, let's be consistent.

Reply:
Files provide `walkFileTree(..)` function to traverse the folders in depth first manner.

Whereas, Location only provides `list()` method to list all the folders together. This can cause high Disk I/O. Location does not provide a function to iterate the file tree.

Reply:
It will still do full list in the end, so there is no much difference in IO. Yes, it does not provide visitor pattern out of the box, but depth first can be simply implemented with recursion or stack. I don't think it's a reason to limit the functionality to local files only.

Reply:
Replaced File references with Location.